### PR TITLE
Spawn tasks

### LINKED ISSUE #5: Add a way to spawn tasks on the local shard in Rust, similarly to tokio::spawn
[`tokio::spawn`](https://docs.rs/tokio/latest/tokio/fn.spawn.html)  is a method that allows to spawn a new asynchronous task in the background.

The task is completely independent - subsequent operations done by the caller are done in parallel and aren't synchronized with the task in any way.

It should be possible to asynchronously wait for the task to finish.
The task might return a result which should also be accessible to the caller.

```rust
async fn main() {
    let task: JoinHandle<i32> = tokio::spawn(long_computation());

    // Compute something in parallel to long_computation()
    let computed: i32 = other_long_computation().await;

    // Wait for the task to finish
    let task_res: i32 = task.await.unwrap();

    let result: i32 = computed + task_res;
}

```

The big difference with `tokio` is that `seastar` has the shard-per-core architecture and each task is executed on a single shard. That means that the future doesn't have to be `Send` or `Sync`. In tokio `Send` and `Sync` are a must because the tasks can be migrated between cores.

Some useful seastar functions:
* `make_task`
* `schedule` - usually isn't used manually but maybe this time it can

### PR DESCRIPTION
**Pull request by sylwiaszunejko**
Fixes: #5 
Depends on: #24, #27 

This PR introduces a way to spawn tasks on the local shard in Rust by method:
```rust
pub async fn spawn<T, Ret: 'static>(future: T) -> Option<Ret>
where
    T: Future<Output = Ret> + 'static
```
For example:
```rust
let _ = spawn(async move {
            spawn(async move {
                return 2;
            })
            .await
            .unwrap()
        })
        .await;
```

### CODE REVIEW

File: seastar/src/spawn.rs [hunk 1]
```diff
@@ -41,3 +41,47 @@ where
...
+    async fn test_spawn_int() {
+        let res = spawn(async move {
+            return 0;
+        })
+        .await;
+        assert!(matches!(res, Some(0)));
+    }
+
+    #[seastar::test]
+    async fn test_two_spawn_int_and_void() {
```
Comment (kfernandez31) [inline seastar/src/spawn.rs:75]: I think we can replace this test with a "`test_spawn_void`" to avoid redundancy.

File: seastar/src/spawn.rs [hunk 2]
```diff
@@ -1,53 +1,43 @@
...
         include!("seastar/src/spawn.hh");
 
-        fn cpp_spawn_void(future: VoidFuture) -> VoidFuture;
-        fn cpp_spawn_int(future: IntFuture) -> IntFuture;
+        fn cpp_spawn(future: VoidFuture) -> VoidFuture;
     }
 }
 
-/// Spawns a new asynchronous task, returning a `VoidFuture`.
+/// Spawns a new asynchronous task, returning a `Option<Ret>`.
```
Comment (kfernandez31) [inline seastar/src/spawn.rs]: Nit: "a" -> "an" :p

**Review by kfernandez31 — CHANGES_REQUESTED** [review]

File: seastar/src/spawn.rs (hunk 1 above)
Comment (sylwiaszunejko) [reply seastar/src/spawn.rs:75]: Do you mean I should delete this test? I wrote it because there was a moment where a single spawn worked, but multiple spawns in a row didn't.

File: seastar/src/spawn.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+/// Spawns a new asynchronous task, returning an `Option<Ret>`.
+///
+/// The provided future will start running in the background immediately
+/// when `spawn` is called.
+///
+/// Spawning a task enables the task to execute concurrently to other tasks.
+///
+/// This function must be called from the context of a Seastar runtime.
+pub async fn spawn<T, Ret: 'static>(future: T) -> Option<Ret>
```
Comment (piodul) [inline seastar/src/spawn.rs]: Why does it return an option? I think it should just return `cxx_async::CxxAsyncResult<Ret>`, like in submit_to.

File: seastar/src/spawn.rs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    T: Future<Output = Ret> + 'static,
+{
+    let x: Rc<Cell<Option<Ret>>> = Default::default();
+    let xcloned = x.clone();
+    cpp_spawn(VoidFuture::infallible_local(async move {
+        xcloned.set(Some(future.await));
+    }))
+    .await
+    .ok()?;
+    Rc::try_unwrap(x).ok().unwrap().into_inner()
```
Comment (piodul) [inline seastar/src/spawn.rs]: This line is very strange. Why `Rc::try_unwrap`?

I would imagine something like this:

```rust
x.take().unwrap()
```

File: seastar/src/lib.rs [hunk 5]
```diff
@@ -8,16 +8,17 @@ mod cxx_async_local_future;
 
 mod config_and_start_seastar;
 mod preempt;
-
```
Comment (piodul) [inline seastar/src/lib.rs:11]: Nit: unnecessary whitespace change

File: seastar/src/spawn.rs [hunk 6]
```diff
@@ -32,6 +32,8 @@ pub async fn spawn<T, Ret: 'static>(future: T) -> Option<Ret>
 where
     T: Future<Output = Ret> + 'static,
 {
+    seastar::assert_runtime_is_running!();
```
Comment (piodul) [inline seastar/src/spawn.rs]: `Ensure safety of spawn`

Please squash this commit with `Add functions to spawn future on the local shard`.

**Comment by piodul** [comment]
Please rebase and drop the commit [Add macro #[seastar::test]](https://github.com/zpp-2022-rust-seastar/seastar-rs/pull/22/commits/b71681613620e47f4460f8f147deba51299be4f8) (the PR for this was merged), as there are conflicts.

**Comment by sylwiaszunejko** [comment]
> Please rebase and drop the commit [Add macro #[seastar::test]](https://github.com/zpp-2022-rust-seastar/seastar-rs/pull/22/commits/b71681613620e47f4460f8f147deba51299be4f8) (the PR for this was merged), as there are conflicts.

done

**Comment by piodul** [comment]
We merged the "API safety" PR, so there are conflicts again... please do the same as before - drop [Add module for checking whether current thread is in a Seastar runtime](https://github.com/zpp-2022-rust-seastar/seastar-rs/pull/22/commits/a664964a26d33b3a5359ccac39cadb850b19a214) and rebase on current main.

Apart from that, I think it looks good to be merged.

File: seastar/build.rs [hunk 7]
```diff
@@ -4,9 +4,11 @@ static CXX_BRIDGES: &[&str] = &[
     // Put all files that contain a cxx::bridge into this list
     "src/preempt.rs",
     "src/config_and_start_seastar.rs",
+    "src/api_safety.rs",
+    "src/spawn.rs",
 ];
 
-static CXX_CPP_SOURCES: &[&str] = &["src/config_and_start_seastar.cc"];
+static CXX_CPP_SOURCES: &[&str] = &["src/config_and_start_seastar.cc", "src/spawn.cc"];
```
Comment (piodul) [inline seastar/build.rs]: One more thing. As each PR will add files to this variable, it would be well to have each file put in a separate line so that we reduce the risk of a merge conflict.

I think a trick used in `CXX_BRIDGES` will prevent rustfmt from putting everything back into a single line: just add a comment similar to the `Put all files...`. Please add it.

File: seastar/src/spawn.hh [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+#include "cxx-async/include/rust/cxx_async_seastar.h"
+#include "rust/cxx.h"
+#include "cxx_async_futures.hh"
+
+namespace seastar_ffi {
+namespace spawn {
+
+VoidFuture cpp_spawn(VoidFuture future);
+
+}
```
Comment (kfernandez31) [inline seastar/src/spawn.hh]: Nit: no `// namespace <namespace name>` comments (it's a good practice to have them)

File: seastar/src/spawn.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+/// This function must be called from the context of a Seastar runtime.
+pub async fn spawn<T, Ret: 'static>(future: T) -> cxx_async::CxxAsyncResult<Ret>
+where
+    T: Future<Output = Ret> + 'static,
+{
+    seastar::assert_runtime_is_running!();
+
+    let x: Rc<Cell<Option<Ret>>> = Default::default();
+    let xcloned = x.clone();
+    match cpp_spawn(VoidFuture::infallible_local(async move {
```
Comment (kfernandez31) [inline seastar/src/spawn.rs:40]: This can be shortened:
```rs
    seastar::assert_runtime_is_running!();

    let x: Rc<Cell<Option<Ret>>> = Default::default();

    let x_clone = x.clone();
    cpp_spawn(VoidFuture::infallible_local(async move {
        x_clone.set(Some(future.await));
    }))
    .await?;

    Ok(x.take().unwrap())
    ```

File: seastar/src/spawn.rs (hunk 1 above)
Comment (kfernandez31) [reply seastar/src/spawn.rs:75]: I see. What bugs me is the asymmetry of the `spawn_int` and `spawn_void` calls. I'd add a second `spawn_int` call in this test and then either keep it as is or split it into two.

Comment (kfernandez31) [reply seastar/src/spawn.rs:75]: You can also think of testing those tests' logic with the `join!` macro - it'd make them more general over the order of the spawns (of course, in these tests, the tasks are kind of identical, so it doesn't matter, but that doesn't have to be the case).

Comment (sylwiaszunejko) [reply seastar/src/spawn.rs:75]: But there are two `spawn_int` calls. I still don't get what the problem is.

File: seastar/src/spawn.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    let x: Rc<Cell<Option<Ret>>> = Default::default();
+
+    let x_clone = x.clone();
+    cpp_spawn(VoidFuture::infallible_local(async move {
+        x_clone.set(Some(future.await));
+    }))
+    .await?;
+
+    Ok(x.take().unwrap())
```
Comment (Lorak-mmk) [inline seastar/src/spawn.rs]: I'm not yet completely convinced about usage of `Rc<Cell<Option<T>>>` - see comment https://github.com/zpp-2022-rust-seastar/seastar-rs/pull/26#discussion_r1125474021 as it applies here too.

Comment (sylwiaszunejko) [reply seastar/src/spawn.rs]: Do you suggest using oneshot::channel? Previously I think @piodul said that it is overkill because the channel will be unnecessarily thread-safe.

Comment (piodul) [reply seastar/src/spawn.rs]: Based on the precedent made here: https://github.com/zpp-2022-rust-seastar/seastar-rs/pull/26#discussion_r1125669863 I think we can keep using `Rc<Cell<Option<T>>>` here as well.

A thread-safe channel is an overkill IMO. The `Rc/Cell/Option` stuff is hidden behind a nice abstraction so this should be fine.

Comment (piodul) [reply seastar/src/spawn.rs]: If this pattern repeats frequently then I'm fine - in the future - with either creating or using (from an external crate) a nicer abstraction for this with desired semantics and performance characteristics.

File: seastar/src/spawn.rs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+/// Spawns a new asynchronous task, returning an `cxx_async::CxxAsyncResult<Ret>`.
+///
+/// The provided future will start running in the background immediately
+/// when `spawn` is called.
+///
+/// Spawning a task enables the task to execute concurrently to other tasks.
+///
+/// This function must be called from the context of a Seastar runtime.
+pub async fn spawn<T, Ret: 'static>(future: T) -> cxx_async::CxxAsyncResult<Ret>
```
Comment (piodul) [inline seastar/src/spawn.rs]: One more thing, related to the conversation on the chat: I don't think it actually makes sense for this function return a Result - it should just return Ret. The only place where a failed result can be generated is `cpp_spawn`, which indicates an exception on the c++ side. I can't imagine any other failure apart from an allocation failure, and Rust just panics/aborts on allocation failure, so I think it makes sense to do the same here - and it simplifies the interface.

**Review by piodul — COMMENTED** [review]
Looks good to me to merge after changing the signature to return `Ret`, not `Result<Ret, _>`.

File: seastar/src/spawn.rs (hunk 11 above)
Comment (Lorak-mmk) [reply seastar/src/spawn.rs]: If we want to use this with Scylla, then don't we need to handle allocation failures?

Comment (piodul) [reply seastar/src/spawn.rs]: Not necessarily. Seastar itself considers some allocations to be irrecoverable. Grep for [scoped_critical_alloc_section](https://github.com/scylladb/seastar/blob/799a9aacb53db4f13c35caecf5071cf078b37159/include/seastar/util/critical_alloc_section.hh#L46-L68) in the Seastar code and see for yourself, it is frequently used in future-related primitives.

Besides that, Rust is ill-prepared to handle allocation failures at this moment of time, unfortunately. Some primitives that allocate (Box, Vec) do have methods that allow handling such failures, but they are opt-in and everything else just aborts - not even panic. We will just have to be super careful not to allocate large chunks of data and use fallible allocation APIs for user data.

Finally, remember that aborting is considered to be safe behavior from Rust's POV.

**Review by piodul — APPROVED** [review]
